rust/gui-client/src-tauri/src/client/gui.rs

```diff
@@ -685,6 +683,16 @@ impl Controller {
                 })?;
                 Ok(())
             }
+            IpcServerMsg::ConnectResult(result) => match result {
```

Comment:
Currently, our IPC design is event-based, i.e. none of the messages have a response that is enforced by the implementation.

I think there are two paths we can go from here:

1. Design our IPC abstraction to enforce a response and generically fail with a timeout to the caller if one isn't provided.
2. Embrace the event style and design messages and client to handle that.

If we go with (2), then I don't think we should name this `ConnectResult`. Instead, I think there should be dedicated event for `TunnelFailed` that gets emitted when `connect` itself fails or the session fails at any point later down the line. It also means we should not code our state machine to await any particular event later down the line in order to resolve a "pending" state.

What I'd like to avoid is assuming that messages have a response but it is not enforced by the IPC abstraction. The problem there is that we it opens us up to these kind of bugs and there is no assistance by the type-system to help us. They are just bugs in a state machine that are really hard to find.

If we go with event-based, we should try and minimise the number of events and be ready to handle all of them in any order as best as possible.

If we go with enforcing request-response, we can build a more rigid state machine in the client because the state transitions are better defined.

Reply:
Yeah. We can't do pure request-response but maybe we can start by separating the request-response pairs from the fire-and-forget events.

Client-initiated requests:
- Connect
- Clear logs

Client-sent events:
- Disconnect
- SetDns
- Reset

Server-sent events:
- OnDisconnect
- OnUpdateResources
- TerminatingGracefully

Worst case, all the requests could be happening concurrently. Right now that's just `Connect` and `ClearLogs` but it would be nice for the IPC service to respond with an ack to the client-sent events just so we know it's not hung or anything.

So the hard part for doing requests properly is that it means more concurrency in the GUI. But yeah we should do it.

Reply:
> So the hard part for doing requests properly is that it means more concurrency in the GUI. But yeah we should do it.

Maybe the other option is easier for now? Don't expect responses and instead:

- Merge `ConnectResult`, `Disconnected` and `TerminatingGracefully`
- Don't expect a response to `ClearLogs` (I think it currently does)

Reply:
I need the response to `ClearLogs` so I know when to re-count the log files

Reply:
> I need the response to `ClearLogs` so I know when to re-count the log files

That could be modeled as its own event that says `IpcLogsUpdated` which always triggers a recount.